android: Ensure we use the local Maven repo to get react-native

gnprice opened the pull request:
## Summary

This fixes #35204.  It ensures that for getting the Android binary dependencies provided by React Native, the build always looks only in the local Maven repo inside node_modules/react-native/.

Without this, the build also looks in the other configured repositories like Maven Central.  Those sometimes have RN binaries in them, by mistake or otherwise, which may be the wrong version.

For how this snippet works, see Gradle documentation:
  https://docs.gradle.org/current/userguide/declaring_repositories.html#sec:repository-content-filtering

Thanks to inckie for pointing out this feature:
  https://github.com/facebook/react-native/issues/35204#issuecomment-1304255623

## Changelog

[Android] [Fixed] - Always get React Native Android libraries from node_modules

## Test Plan

I applied this change to my own project where #35204 was reproducing, and the build started succeeding again.

mikehardy commented:
@cortinico + @brentvatne you may both be interested in this - it came from the collaboration on the github issue related to the android build break today

gnprice commented:
CI is failing. It looks like the issue is that the CI job relies on passing an extra Gradle property `REACT_NATIVE_MAVEN_LOCAL_REPO` to specify where to look for the local Maven repo:
```
# .circleci/config.yml
            ./gradlew assemble<< parameters.flavor >> -PREACT_NATIVE_MAVEN_LOCAL_REPO=/root/react-native/maven-local
```

Then the way that property is given meaning is this bit in the RN Gradle plugin, in `packages/react-native-gradle-plugin/src/main/kotlin/com/facebook/react/utils/DependencyUtils.kt`:
```kotlin
  fun configureRepositories(project: Project, reactNativeDir: File) {
    with(project) {
      if (hasProperty("REACT_NATIVE_MAVEN_LOCAL_REPO")) {
        mavenRepoFromUrl("file://${property("REACT_NATIVE_MAVEN_LOCAL_REPO")}")
      }
```
That is, when the property is present, the plugin configures an additional Maven repo based on it. That gets defeated when there's already an `exclusiveContent` pointing elsewhere for the desired content, as there is with this PR.

I think it should be straightforward to fix things so that this fix for #35204 works without interfering with the use of the `REACT_NATIVE_MAVEN_LOCAL_REPO` property -- the simplest way is probably to adjust the template `app/build.gradle` code so that it too respects that property. I'll take a look at that sometime in the next few days.

Perhaps better: it seems like that code in the RN Gradle plugin is already doing a job that overlaps with that code in the template `app/build.gradle` script. Perhaps that logic can be deduplicated, simplifying the script? That'd be neat but possibly out of scope for fixing the present issue. I might take a look at that after I get a working fix that passes CI.

cortinico commented:
Hey @gnprice
Thanks for sending this PR.

This change as it is, it's actually unnecessary as it will land on main (so wilk be shipped on 0.72).

You should send this same PR agains the 0.68-stable, 0.69-stable and 0.70-stable.

0.71 is instead unaffected as it will consume artifacts from Maven Central. That's also the reason why the CI is red.

cortinico commented:
Closing for the aforemntioned reasons + we provided hotfixes for the last 8 stable releases

gnprice commented:
> 0.71 is instead unaffected as it will consume artifacts from Maven Central

Ah, and I see the explanation that's now on #35210:
> In 0.71 we cleaned up the new app template and removed all the `+` dependencies ([see here](https://github.com/facebook/react-native/blob/f0b7cbe22e2d67c102149d6d118b563b71a7fbb5/template/android/app/build.gradle#L152-L153)) and we moved the template to use the React Native Gradle Plugin, which will allow us to prevent scenarios like this from happening in the future.

That's great news! Glad this is already fixed thoroughly in main, and thanks for providing those stable-version fixes.
